internal/compiler/wasm/wasm.go

```diff
@@ -635,15 +628,37 @@ func (c *Compiler) compileFunc(fn *ir.Func) error {
 
 	c.code = &module.CodeEntry{}
 
+	// memoization: get
+	if memoize {
+		c.appendInstr(instruction.Block{Instrs: []instruction.Instruction{
+			instruction.I32Const{Value: int32(idx)},
+			instruction.Call{Index: c.function(opaMemoizeGet)},
+			instruction.TeeLocal{Index: c.local(fn.Return)},
+			instruction.I32Eqz{},
+			instruction.BrIf{Index: 0},
+			instruction.GetLocal{Index: c.local(fn.Return)},
+			instruction.Return{},
+		}})
+	}
+
 	for i := range fn.Blocks {
 		instrs, err := c.compileBlock(fn.Blocks[i])
 		if err != nil {
 			return errors.Wrapf(err, "block %d", i)
 		}
-		if i < len(fn.Blocks)-1 {
+		if i < len(fn.Blocks)-1 { // not the last block: wrap in `block` instr
 			c.appendInstr(instruction.Block{Instrs: instrs})
-		} else {
-			c.appendInstrs(instrs)
+		} else { // last block, no wrapping
+			// memoization: insert, spliced into the instructions right
+			// before the return:
+			for _, instr := range instrs {
+				if _, ok := instr.(instruction.Return); ok && memoize {
+					c.appendInstr(instruction.I32Const{Value: int32(idx)})
+					c.appendInstr(instruction.GetLocal{Index: c.local(fn.Return)})
+					c.appendInstr(instruction.Call{Index: c.function(opaMemoizeInsert)})
+				}
+				c.appendInstr(instr)
```

Comment:
💭 Currently, if this instruction **is** `Return`, we could just ignore it. The last value on the stack is the function's return value, with our without that statement.